bump(main/python): 3.13

thunder-coding opened the pull request:
Mostly ready for review, waiting for CI to turn green and some testing to ensure nothing is broken. Every change is in `python-fullrebuild` branch, so it is recommend to review that branch. Branch `python313` as well as `python313-dep{1,2,...}` branches are derived from the `python-fullrebuild` branch. The list of packages that needed to be rebuilt was generated using `apt-file search lib/python3.12`. Those packages were than later replaced with their parent package names (in case of subpackages), and then some scripts were used to intelligently sort the packages into buckets for rebuild. The list of packages was then saved to `output/rebuildlist.txt`

Script used to generate rebuild list. This script was specifically designed to take into account large packages that needed more disk space for rebuild as well as packages that depended on python and also were dependencies of other python packages:

```sh
#!/usr/bin/bash
set -eou pipefail

rm -f output/rebuildlist-*.txt

declare -a packages=()
while IFS= read -r package; do
	packages+=("$package")
done < output/rebuildlist.txt

orig_packages=("${packages[@]}")
orig_grep_str=""
for pkg in "${orig_packages[@]}"; do
	orig_grep_str+="|$pkg"
done
orig_grep_str="${orig_grep_str:1}"

current_index=1

_generate_rebuild_list() {
	declare -a new_packages=()
	declare -a curr_packages=()
	local grep_str=""
	local -a args=("$@")
	for ((i = 0; i < $#; i++)); do
		pkg="${args[$i]}"
		grep_str+="|$pkg"
	done
	grep_str="${grep_str:1}"
	# echo "DEBUG: Grep string is ($grep_str)"
	for package in "${packages[@]}"; do
		local extra_dirs=()
		if [[ ! -f "packages/$package/build.sh" ]]; then
			extra_dirs+=("packages")
		elif [[ ! -f "x11-packages/$package/build.sh" ]]; then
			extra_dirs+=("x11-packages")
		elif [[ ! -f "root-packages/$package/build.sh" ]]; then
			extra_dirs+=("root-packages")
		fi
		for pkg in "$@"; do
			if [[ "$pkg" == "$package" ]]; then
				curr_packages+=("$package")
				continue 2
			fi
		done
		if ./scripts/buildorder.py ./*packages/"$package" "${extra_dirs[@]}" | grep -qE "($grep_str)$"; then
			curr_packages+=("$package")
			continue
		fi
		new_packages+=("$package")
	done

	local rebuild_list=()
	for pkg in "${curr_packages[@]}"; do
		local extra_dirs=()
		if [[ ! -f "packages/$pkg/build.sh" ]]; then
			extra_dirs+=("packages")
		elif [[ ! -f "x11-packages/$pkg/build.sh" ]]; then
			extra_dirs+=("x11-packages")
		elif [[ ! -f "root-packages/$pkg/build.sh" ]]; then
			extra_dirs+=("root-packages")
		fi
		for rebuild_pkg in $(./scripts/buildorder.py ./*packages/"$pkg" "${extra_dirs[@]}" | grep -oE "($orig_grep_str)$"); do
			rebuild_list+=("$rebuild_pkg")
		done
		rebuild_list+=("$pkg")
	done
	# Remove duplicates from rebuild_list
	rebuild_list=($(echo "${rebuild_list[@]}" | tr ' ' '\n' | sort -u | tr '\n' ' '))

	echo "DEBUG: Dealt with $(( ${#packages[@]} - ${#new_packages[@]} )) packages, ${#new_packages[@]} remain. This step 
```


Script to generate `python313-dep*` branches:

```sh
#!/usr/bin/bash
set -eou pipefail

rm -f output/rebuildlist-*.txt

declare -a packages=()
while IFS= read -r package; do
	packages+=("$package")
done < output/rebuildlist.txt

orig_packages=("${packages[@]}")
orig_grep_str=""
for pkg in "${orig_packages[@]}"; do
	orig_grep_str+="|$pkg"
done
orig_grep_str="${orig_grep_str:1}"

current_index=1

_generate_rebuild_list() {
	declare -a new_packages=()
	declare -a curr_packages=()
	local grep_str=""
	local -a args=("$@")
	for ((i = 0; i < $#; i++)); do
		pkg="${args[$i]}"
		grep_str+="|$pkg"
	done
	grep_str="${grep_str:1}"
	# echo "DEBUG: Grep string is ($grep_str)"
	for package in "${packages[@]}"; do
		local extra_dirs=()
		if [[ ! -f "packages/$package/build.sh" ]]; then
			extra_dirs+=("packages")
		elif [[ ! -f "x11-packages/$package/build.sh" ]]; then
			extra_dirs+=("x11-packages")
		elif [[ ! -f "root-packages/$package/build.sh" ]]; then
			extra_dirs+=("root-packages")
		fi
		for pkg in "$@"; do
			if [[ "$pkg" == "$package" ]]; then
				curr_packages+=("$package")
				continue 2
			fi
		done
		if ./scripts/buildorder.py ./*packages/"$package" "${extra_dirs[@]}" | grep -qE "($grep_str)$"; then
			curr_packages+=("$package")
			continue
		fi
		new_packages+=("$package")
	done

	local rebuild_list=()
	for pkg in "${curr_packages[@]}"; do
		local extra_dirs=()
		if [[ ! -f "packages/$pkg/build.sh" ]]; then
			extra_dirs+=("packages")
		elif [[ ! -f "x11-packages/$pkg/build.sh" ]]; then
			extra_dirs+=("x11-packages")
		elif [[ ! -f "root-packages/$pkg/build.sh" ]]; then
			extra_dirs+=("root-packages")
		fi
		for rebuild_pkg in $(./scripts/buildorder.py ./*packages/"$pkg" "${extra_dirs[@]}" | grep -oE "($orig_grep_str)$"); do
			rebuild_list+=("$rebuild_pkg")
		done
		rebuild_list+=("$pkg")
	done
	# Remove duplicates from rebuild_list
	rebuild_list=($(echo "${rebuild_list[@]}" | tr ' ' '\n' | sort -u | tr '\n' ' '))

	echo "DEBUG: Dealt with $(( ${#packages[@]} - ${#new_packages[@]} )) packages, ${#new_packages[@]} remain. This step rebuilds $((${#rebuild_list[@]} - ${#curr_packages[@]})) extra packages."

	for pkg in "${curr_packages[@]}"; do
		echo "$pkg"
	done > output/rebuildlist-$current_index.txt
	for pkg in "${rebuild_list[@]}"; do
		for curr_pkg in "${curr_packages[@]}"; do
			if [[ "$pkg" == "$curr_pkg" ]]; then
				continue 2
			fi
		done
		echo "$pkg"
	done > output/rebuildlist-$current_index-extra.txt

	packages=("${new_packages[@]}")
	((current_index++))
}
_generate_rebuild_list libllvm flang python-llvmlite python-libsass xapp asciidoc
_generate_rebuild_list python-torch pyqt5
_generate_rebuild_list pygobject pycairo opencv dbus-python python-lxml
_generate_rebuild_list libxml2
_generate_rebuild_list python-numpy python-greenlet python-msgpack python-brotli python-pillow python-scipy
_generate_rebuild_list boost 2ping borgbackup fetchmail speechd vapoursynth
_generate_rebuild_list python-cryptography python-skia-pathops python-pycryptodomex libunbound libsearpc
_generate_rebuild_list "${packages[@]}"
```

Only the `python313` branch needs to be generated full from scratch before running the above script


> [!WARNING]
> As of writing this. the list below (list of todos along with rebuild list) is of the local changes I have, and may not be in sync with the branches and PRs on github.

The following is the significant part that needs to be done first:
- [X] Move away from Ubuntu's python when building cross python. We need to bootstrap same version of build-python first. Thankfully I had done this a few days ago for an unrelated project, so I have the build scripts with me for an older version that works, I just need to port it for Python 3.13
- [X] Fix ctypes builtin package not detecting packages for cross-python. `ctypes.utils.find_package()` isn't working. (Tracked down to python/cpython@1f8b24ef69896680d6ba6005e75e1cc79a744f9e)
- [X] Get all things building (tested locally for aarch64. Waiting for CI to turn green)
- [X] Get pip packages to build
- [ ] Full test suite locally on Termux + termux-docker, like I do for nodejs


> [!CAUTION]
> This is a huge **BREAKING CHANGE**. Seems like upstream changed a lot of stuff which is going to break a lot of stuff for users. Major breaking changes which I found will affect us most:

> [!NOTE]
> Since nobody seems to have started this work since a long time, I am also thinking of taking up maintainership of python. Will try to keep this up to date with full test coverage.


<details>
  <summary>`sys.platform` now reports `"android"`</summary>

Earlier versions of uptil 3.12 used to have `sys.platform` to `"linux"`

```python
import sys

# Prints "android" on Python >= 3.13
# and    "linux"   on Python <= 3.12
print(sys.platform)
```

This will very likely break scripts that assume Android to just be Linux earlier. Might need new changes in packages downstream. This is going to be definitely a lot of fun to deal with. Although shouldn't be that much pain considering Android is now a supported platform
</details>


<details>
<summary> Rebuild Step 1 #27768 </summary>
List of packages that will be rebuilt:

```
asciidoc
cinnamon
codelldb
flang
libllvm
python-libsass
python-llvmlite
python-xapp
termux-apt-repo
xapp
```

List of extra packages that need to be rebuilt 1 (due to extra dependencies):
```
at-spi2-core
dbus-python
libunbound
libxml2
pycairo
pygobject
python-lxml
python-pillow
xcb-proto
```
</details>


<details>
<summary> Rebuild Step 2 #27777 </summary>
List of packages that will be rebuilt:

```
pyqt5
python-pyqtwebengine
python-qscintilla
python-torch
python-torchaudio
python-torchcodec
python-torchvision
```

List of extra packages that need to be rebuilt 2 (due to extra dependencies):
```
at-spi2-core
boost
libunbound
libxml2
python-lxml
python-numpy
python-pillow
xcb-proto
```
</details>


<details>
<summary> Rebuild Step 3 #27977 </summary>
List of packages that will be rebuilt:

```
blueprint-compiler
dbus-python
gjots2
gnumeric
gst-python
libncnn
manim
meld
mlt
opencv
orca
pyatspi
pycairo
pygobject
python-lxml
python-opencv-python
```

List of extra packages that need to be rebuilt 3 (due to extra dependencies):
```
at-spi2-core
boost
libunbound
libxml2
python-numpy
python-pillow
python-scipy
python-skia-pathops
speechd
xcb-proto
```
</details>


<details>
<summary> Rebuild Step 4 #28180 </summary>
List of packages that will be rebuilt:

```
at-spi2-core
gnucash
libxml2
nala
python-apt
python-cmake
recoll
```

List of extra packages that need to be rebuilt 4 (due to extra dependencies):
```
boost
libunbound
xcb-proto
```
</details>


<details>
<summary> Rebuild Step 5 #28181 </summary>
List of packages that will be rebuilt:

```
libarrow-cpp
matplotlib
python-brotli
python-contourpy
python-greenlet
python-msgpack
python-numpy
python-onnxruntime
python-pillow
python-pynvim
python-scipy
python-tflite-runtime
python-yt-dlp
```

List of extra packages that need to be rebuilt 5 (due to extra dependencies):
```
boost
python-pycryptodomex
xcb-proto
```
</details>


<details>
<summary> Rebuild Step 6 #28182 </summary>
List of packages that will be rebuilt:

```
2ping
boost
borgbackup
fetchmail
ledger
libtorrent-rasterbar
speechd
tizonia
vapoursynth
```

List of extra packages that need to be rebuilt 6 (due to extra dependencies):
```
xcb-proto
```
</details>


<details>
<summary> Rebuild Step 7 #28183 </summary>
List of packages that will be rebuilt:

```
electrum
libsearpc
libunbound
python-cryptography
python-pycryptodomex
python-skia-pathops
seafile-client
```

List of extra packages that need to be rebuilt 7 (due to extra dependencies):
```
xcb-proto
```
</details>


<details>
<summary> Rebuild Step 8 #28184 </summary>
List of packages that will be rebuilt:

```
frida
python-apsw
python-bcrypt
python-grpcio
python-lameenc
python-sabyenc3
python-tldp
python-xlib
ranger
rdiff-backup
ruff
udocker
xcb-proto
yt-dlp-ejs
```

List of extra packages that need to be rebuilt 8 (due to extra dependencies):
```
```
</details>

thunder-coding committed:
bump(main/python): 3.13.11

patch for configure is no longer needed

patch for multiprocessing.c no longer required due to
python/cpython@fa1d675309c6a08b0833cf25cffe476c6166aba3

Remove 0006-do-not-use-xattr.patch since we can just add a configure
argument and it works just as well

We no longer need to manually add libraries for libpython as it is now
added by default when building for android. See upstream commit:
python/cpython@7f5e3f04f838686d65f1053a5e47f5d3faf0b228

No longer required patch for Lib/aifc.py after
python/cpython@036da3bd43aa2593d17d2fb73d4794f9965c577d

mailcap module has been removed in Python 3.13.
https://docs.python.org/3/whatsnew/3.13.html#whatsnew313-pep594

add post install message when site-packages/ is detected for python3.12

Also mark as unsafe for on-device building. The plan is to move to building
python with host-python, this isn't possible on device. I will be
looking to get some fixes made in future to get on-device builds as
well, but it is not the priority at the moment

thunder-coding committed:
chore(main/python): build with parallelism [[TESTING REQUIRED]]

maybe the issue with the build system is fixed now?

robertkirkman commented on packages/python/build.sh:
```diff
@@ -17,9 +17,6 @@ TERMUX_PKG_REPLACES="python-dev"
 TERMUX_PKG_PROVIDES="python3"
 
-# https://github.com/termux/termux-packages/issues/15908
-TERMUX_PKG_MAKE_PROCESSES=1
```
For convenient reference, here is the series of commands which need to work consistently (no matter how many times they are repeated, even if repeated **many times**) when this is removed:

```bash
ssh -p 8022 192.168.12.185 # termux device ip address
rm -rf output
exit
rm -rf output
scripts/run-docker.sh ./build-package.sh -I -f python
scp -r -P 8022 output/ 192.168.12.185:~
ssh -p 8022 192.168.12.185
cd output/
apt reinstall ./*.deb
python -c "import _random"
exit
```

thunder-coding commented:
> Is there any specific reason this isn't the newest version, [3.14.2](https://www.python.org/downloads/release/python-3142/)?

I assume that python 3.14 might have some breaking changes that might not be suitable for everyone yet. Will upgrade to 3.13 now, then will target 3.14 in the next few months. I'm targeting python 3.13 by mid-January.

thunder-coding committed:
scripts(termux_setup_build_python): build-python for cross compilation

We need build python for cross compilation as Ubuntu's version of python
is different. Also crossenv recommends building own python revisions
instead of distribution compiled ones

added lld to setup-ubuntu as it's needed by clang driver when using
-fuse-ld=lld. Not sure how we went this far without having to deal with
this during host builds

Also we need libpython3.so for cross-compilation of some of the pip
packages. So build with --enable-shared

thunder-coding committed:
scripts(termux_setup_python_pip): use build python and add minor
documentation

BREAKING CHANGE: This potentially breaks a lot of builds that might need
python with ubuntu's packaged pip packages/additional features

thunder-coding committed:
rebuild(main/xcb-proto): with python3.13

thunder-coding committed:
rebuild(main/2ping): with python3.13

thunder-coding commented on packages/python/build.sh:
```diff
@@ -16,9 +16,8 @@ TERMUX_PKG_BREAKS="python2 (<= 2.7.15), python-dev"
-# https://github.com/termux/termux-packages/issues/15908
-TERMUX_PKG_MAKE_PROCESSES=1
+# Python build is a 2-step process. Requiring host build and cross build
+TERMUX_PKG_ON_DEVICE_BUILD_NOT_SUPPORTED=false
```
Currently my priority is not on-device builds. I want to first fix docker builds. I'll be reverting this once testing is done on-device and on termux-docker for CI builds of python. And that was supposed to be `TERMUX_PKG_ON_DEVICE_BUILD_NOT_SUPPORTED=true`

robertkirkman replied:
Note also:

- `false` is the default value of `TERMUX_PKG_ON_DEVICE_BUILD_NOT_SUPPORTED`, so this line as written doesn't do anything, which might not be what you were initially intending
- `TERMUX_PKG_ON_DEVICE_BUILD_NOT_SUPPORTED=true` is intended specifically for packages that edit files during the build that will _not be tracked by the same package_ in the package manager, for example, `qt6-qtwebengine` is `TERMUX_PKG_ON_DEVICE_BUILD_NOT_SUPPORTED=true` because it edits the file `$TERMUX_PREFIX/lib/librt.so` during the build, which is a file normally tracked by the package `ndk-sysroot`. Does that situation apply to this `python` in some way?


 https://github.com/termux/termux-packages/blob/b9da78bedecdc76ba405c2a821a6e5a2d821bb6c/x11-packages/qt6-qtwebengine/build.sh#L106

thunder-coding committed:
rebuild(main/python-pip): with python3.13

robertkirkman commented on packages/python-pip/build.sh:
```diff
@@ -3,6 +3,7 @@ TERMUX_PKG_DESCRIPTION="The PyPA recommended tool for installing Python packages
 TERMUX_PKG_LICENSE="MIT"
 TERMUX_PKG_MAINTAINER="@termux"
 TERMUX_PKG_VERSION="25.3"
+TERMUX_PKG_REVISION=1
```
Unfortunately, the `python-pip` package built here is broken with this error at runtime:

```bash
/data/data/com.termux/files/usr/bin/pip: 2: exec: /home/builder/.termux-build/python3.13-crossenv-prefix-bionic-x86_64/cross/bin/python3.13: not found
```

And I have determined that this error is caused by this file

```
/home/builder/.termux-build/python3.13-crossenv-prefix-bionic-x86_64/build/lib/python3.13/site-packages/pip/_vendor/distlib/scripts.py
```

being bumped past a version that now contains this iOS-related PR:

- https://github.com/pypa/distlib/pull/231
- https://github.com/pypa/pip/pull/12962
- https://github.com/pypa/pip/commit/c28facaadcf92ace8ed098245b613e6d1bd98304

(specifically this part)

<img width="918" height="303" alt="image" src="https://github.com/user-attachments/assets/63ffd23d-214b-4c09-93ab-970d0024fe77" />


There are several potential ways I can think of to work around this going forward:

- One option is to apply a patch to that file right after it's created (it's created during `termux_setup_python_pip()`) that simply reverts the entire change that causes the problem to start happening
- Another option is to try to make a patch with a new, unique change in the same file that works around the new `pip` code in some way that doesn't involve completely reverting it
- Another option is to try to change this section of `termux_step_massage()` (which is tangentially involved) to handle the problem in a similar way to how it handles the currently-working codepath, but adapted to also handle the new codepath that upstream `pip` switched to without reverting it

https://github.com/termux/termux-packages/blob/b9da78bedecdc76ba405c2a821a6e5a2d821bb6c/scripts/build/termux_step_massage.sh#L98-L113

If I were to work on this I'd probably just apply a patch that reverts the upstream change if doing so allows everything in Termux to keep working, but you can try whatever seems best to you and I hope that this summary of the problem and possible options helps you somewhat.

robertkirkman commented:
### Python 3.13 [Upstream Android support (PEP738)](https://peps.python.org/pep-0738/) test notes

- Appears to be working successfully (after working around the issue involving `pip` described above)
- `pip install` may download **native libraries prebuilt for Android** from some PyPi packages rather than building them all from source
- Whether some may include ADB-isms (like attempts to access absolute path `/data/local/tmp`), APK-isms (like attempts to access `ANativeActivity` or `ANativeWindow`), or Android-8-isms (like binaries built targeting a higher API level than 24) remains unknown, but no cases of any of those have yet been observed
- PyPi packages containing native libraries prebuilt for Android identified using this list: https://beeware.org/mobile-wheels/

#### Test 1: `pip install pypdfium2`

- Worked without errors
- **Important:** testing indicates that `pypdfium2` has implemented **Android native library support compatible with Python 3.12**, meaning that **some, but not all, PyPi packages already contain Android native libraries that are successfully deployed to Termux currently, without necessarily requiring Python 3.13**. Because of that, the behavior of `pip install pypdfium2` is the same regardless of whether this PR is applied.

#### Test 2: `pip install optree`

- Worked without errors
- Testing indicates that, unlike the case of `pypdfium2`, `optree` **leverages PEP738 to deploy its Android native library, meaning that `pip install optree` does behave differently in Python 3.13 from how it does in Python 3.12**.
- `Building wheel for optree` did not print; instead, **`Downloading optree-0.18.0-cp313-cp313-android_24_x86_64.whl`** printed
- A `$PREFIX/lib/python3.13/site-packages/optree/_C.cpython-313-x86_64-linux-android.so` was installed
- `file $PREFIX/lib/python3.13/site-packages/optree/_C.cpython-313-x86_64-linux-android.so` prints this:
  - `ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, for Android 24, built by NDK r27d (13750724), BuildID[sha1]=84f99286f68a16ce87232652f5558339de6ddb65, stripped`
- Since it was built by NDK r27d, not NDK r28 or newer, this suggests that a possible side effect could be **incompatibility of some of these PEP738 wheels with 16KB-page-size-only devices**.
- `optree` basic program tested successfully:

```python
#!/usr/bin/env python3

import optree

tree = {'b': (2, [3, 4]), 'a': 1, 'c': 5, 'd': 6}
optree.tree_flatten(tree)
```


#### Summary

- Appears to be working, no Termux-specific interventions appear to be necessary at this time, unless a particular PyPi package is found to be bugged
- Some PyPi packages may temporarily lose compatibility with 16KB-page-size-only devices until their respective upstreams upgrade to NDK r28 or newer; this seems minor, since there are very few 16KB-page-size-only devices.

thunder-coding commented:
Thanks for the brief. I havent started testing yet. Just pushed the changed I had for proper cross-compilation I has been holding since some time. Will be busy this month with some stuff. Wont be moving ahead without running tests for Python

thunder-coding committed:
scripts(termux_setup_meson): use ubuntu's packaged python

Do not try to use build python for running meson

thunder-coding committed:
scripts(termux_setup_python_pip): fix install shebang with crossenv

This is a slightly modified version of
robotpy/crossenv@3ef761b02c45546edcb40ed038f459ce38d29e6b

We only need the simple shebang, so only use that always.

robertkirkman commented on scripts/build/setup/termux_setup_build_python.sh:
```diff
@@ -0,0 +1,80 @@
+				make -j "$(nproc)" install
+			cd ..
+			# Add our own built python to path
+			export PATH="$_PYTHON_FOLDER/host-build-prefix/bin:$PATH"
```
Oh I'm sorry I forgot to tell you that this line needs to be outside of the `if [[ ! -d "$_PYTHON_FOLDER" ]]; then` block otherwise there will be an error when repeated builds are performed

I noticed that yesterday and I thought I told you but I forgot to

thunder-coding committed:
chore(main/python): take up maintainership

thunder-coding committed:
scripts(termux_setup_python_pip): crossenv 1.6.1

Also update setuptools to 80.9.0 while we are at full rebuilds with
python 3.13

thunder-coding committed:
pip hardcode api level instead of getting it from runtime

fixes crossenv broken

thunder-coding commented on scripts/build/setup/python-crossenv-pip-android-api-level.patch:
```diff
@@ -0,0 +1,47 @@
++    if api_level is None:
++        # Python 3.13 was the first version to return platform.system() == "Android",
++        # and also the first version to define platform.android_ver().
++        api_level = @@TERMUX_PKG_API_LEVEL@@  # type: ignore[attr-defined]
```
This patch fixed `cross-pip` throwing up with python 3.13. This should get all python packages built properly

thunder-coding committed:
rebuild(main/borgbackup): with python 3.13

Properly build with cross-pip instead of building using build-python and
installing it. Seems like pip now treats android as a separate platform
than other linux, hence the incompatibility

thunder-coding committed:
fix(main/python): unexpanded autotools variable in pkgconfig

thunder-coding committed:
rebuild(main/dbus-python): with python 3.13

thunder-coding committed:
rebuild(main/libxml2): with python 3.13

The libxml2-python subpackage needs to be rebuilt

thunder-coding committed:
rebuild(main/gobject-introspection): with python 3.13

g-ir-scannner subpackage needs to be rebuilt

thunder-coding committed:
rebuild(main/libplacebo): with python 3.13

use build-python for generation of gl.h file since we want that.

thunder-coding committed:
rebuild(main/libtorrent-rasterbar): with python 3.13

thunder-coding committed:
fix(main/python): ctypes.util.find_library on android

Relevant upstream commit https://github.com/python/cpython/commit/1f8b24ef69896680d6ba6005e75e1cc79a744f9e

Also apply this same fix in host build of python as cross-python uses
it's own cutils instead of what we have built for target in our cross
builds.

thunder-coding committed:
rebuild(main/fetchmail): with python 3.13

thunder-coding committed:
rebuild(x11/kitty): with python 3.13

thunder-coding committed:
rebuild(main/boost): with python 3.13

Needed to fix hardcoded python version in find_package for Boost.Python.
Needed to fix builds for ledger

thunder-coding committed:
rebuild(main/ledger): with python 3.13

thunder-coding committed:
rebuild(main/pybind11): with python 3.13

thunder-coding committed:
rebuild(main/pycairo): with python 3.13

thunder-coding committed:
rebuild(main/python-apsw): with python 3.13

thunder-coding committed:
rebuild(main/python-apt): with python 3.13

thunder-coding committed:
rebuild(main/panda3d): with python 3.13

thunder-coding committed:
scripts(termux_setup_python_pip): allow installing wheels without android api version

`python -m build` builds android wheels as android_$TERMUX_ARCH.whl but
pip during installation expects it to be
android_$TERMUX_API_LEVEL_$TERMUX_ARCH.whl so we need make sure that it
is installable without renaming the wheel

thunder-coding committed:
rebuild(main/python-lameenc): with python 3.13

thunder-coding committed:
rebuild(main/python-libsass): with python 3.13

thunder-coding committed:
rebuild(main/python-msgpack): with python 3.13

thunder-coding committed:
rebuild(main/python-numpy): with python 3.13

thunder-coding committed:
rebuild(main/python-onnxruntime): with python 3.13

thunder-coding committed:
rebuild(main/python-sabyenc3): with python 3.13

thunder-coding committed:
rebuild(main/python-skia-pathops): with python 3.13

thunder-coding committed:
rebuild(main/ranger): with python 3.13

thunder-coding committed:
rebuild(main/speechd): with python 3.13

thunder-coding committed:
rebuild(main/udocker): with python 3.13

thunder-coding committed:
rebuild(main/vapoursynth): with python 3.13

thunder-coding committed:
rebuild(x11/xkeyboard-config): with python 3.13

thunder-coding committed:
rebuild(main/yt-dlp-ejs): with python 3.13

Also make sure we always use npm instead of using pnpm.

The hatch build process tries to build the bundle using pnpm, npm asks
for confirmation, which doesn't work with interactive terminal. CI
doesn't have stdin so it isn't a problem over there.

thunder-coding committed:
fix(main/panda3d): do not override android api level

This overrides the api level thus causing clang driver to pick up
incorrect as binary for building. This went unnoticed as I had only
tested builds for aarch64 locally. The CI picked this up for other archs

thunder-coding committed:
fix(main/panda3d): combine all patches for makepanda.py

thunder-coding committed:
fix(main/python): build without -Wl,--as-needed

Causes the built libpython3.so to not include symbols which we need in
the final binary

thunder-coding committed:
rebuild(main/libsearpc): with python 3.13

thunder-coding committed:
rebuild(main/libunbound): with python 3.13

thunder-coding committed:
rebuild(main/python-bcrypt): with python 3.13

thunder-coding committed:
rebuild(main/python-greenlet): with python 3.13

thunder-coding committed:
rebuild(main/python-pillow): with python 3.13

thunder-coding committed:
rebuild(main/python-grpcio): with python 3.13

thunder-coding committed:
rebuild(main/rdiff-backup): with python 3.13

thunder-coding committed:
rebuild(x11/libncnn): with python 3.13

thunder-coding committed:
fix(main/python): module linking to libpython3.so

thunder-coding committed:
fix(main/panda3d): do not pass --target= as it confuses compiler driver

Also do parallel builds

thunder-coding committed:
fix(main/panda3d): architecture detection for i686

thunder-coding committed:
rebuild(main/asciidoc): with python 3.13

thunder-coding committed:
rebuild(root/frida): with python3.13

use build python for python interpretor. Although I am the guy who has
marked this as not the recommended way to do things, I'm breaking my own
rule here to ensure that we don't have to fuck around the entire git
submodule based project to figure out where to make it properly support
different versions of python for host and target

Also seems like workaround for installation of python site-packages is
no longer needed, so get rid of it as well

thunder-coding committed:
rebuild(main/python-brotli): with python 3.13

thunder-coding committed:
rebuild(main/python-cmake): with python 3.13

thunder-coding committed:
rebuild(main/python-cryptography): with python 3.13

thunder-coding committed:
rebuild(main/python-llvmlite): with python 3.13

thunder-coding committed:
rebuild(main/python-pycryptodomex): with python 3.13

thunder-coding committed:
rebuild(main/python-tldp): with python 3.13

thunder-coding committed:
rebuild(main/python-xlib): with python 3.13

thunder-coding committed:
rebuild(x11/pyqt5): with python 3.13

thunder-coding committed:
rebuild(main/ruff): with python 3.13

ANDROID_API_LEVEL seems to be required for newer maturin versions. Also
set other environment variables used by maturin while we are at it

thunder-coding committed:
rebuild(main/electrum): with python 3.13

thunder-coding committed:
rebuild(x11/gnucash): with python 3.13

thunder-coding committed:
rebuild(main/libarrow-cpp): rebuild with python 3.13

Also use ninja for building pyarrow for faster builds than unix
makefiles. Also remove logic for unsetting PYTHONPATH, as it is needed
for Cython to be used.

thunder-coding committed:
rebuild(main/libllvm): with python 3.13

lldb package has python bindings that need to be rebuilt with python
update. Also remove PYTHON_EXECUTABLE passed to CMake, as it's not used
during generation

thunder-coding committed:
rebuild(main/flang): to sync with libllvm version

thunder-coding committed:
rebuild(main/nala): with python 3.13

thunder-coding committed:
rebuild(x11/opencv): with python 3.13

thunder-coding committed:
rebuild(main/pygobject): with python 3.13

thunder-coding committed:
rebuild(main/python-contourpy): with python 3.13

thunder-coding committed:
rebuild(main/python-pynvim): with python 3.13

thunder-coding committed:
rebuild(main/python-lxml): with python 3.13

thunder-coding committed:
rebuild(x11/python-pyqtwebengine): rebuild with python 3.13

thunder-coding committed:
rebuild(x11/python-qscintilla): with python 3.13

thunder-coding committed:
rebuild(main/python-tflite-runtime): with python 3.13

thunder-coding committed:
rebuild(main/python-scipy): with python 3.13

thunder-coding committed:
rebuild(main/recoll): with python 3.13

thunder-coding committed:
rebuild(main/seafile-client): with python 3.13

thunder-coding committed:
rebuild(main/python-yt-dlp): with python 3.13

thunder-coding committed:
rebuild(main/python-torch): with python 3.13

thunder-coding committed:
rebuild(x11/gjots2): with python 3.13

thunder-coding committed:
rebuild(x11/gnumeric): with python 3.13

thunder-coding committed:
rebuild(main/gst-python): with python 3.13

thunder-coding committed:
rebuild(x11/meld): with python 3.13

thunder-coding committed:
rebuild(x11/pyatspi): with python 3.13

thunder-coding committed:
rebuild(x11/codelldb): with python 3.13

thunder-coding committed:
rebuild(main/python-torchcodec): with python 3.13

thunder-coding committed:
rebuild(main/python-torchvision): with python 3.13

thunder-coding committed:
rebuild(main/manim): with python 3.13

thunder-coding committed:
rebuild(main/matplotlib): with python 3.13

thunder-coding committed:
rebuild(x11/python-opencv-python): rebuild with python 3.13

thunder-coding committed:
rebuild(x11/orca): with python 3.13

thunder-coding committed:
rebuild(x11/python-xapp): with python 3.13

thunder-coding committed:
rebuild(main/python-torchaudio): with python 3.13

thunder-coding committed:
rebuild(main/at-spi2-core): with python 3.13

thunder-coding committed:
rebuild(main/blueprint-compiler): with python 3.13

thunder-coding committed:
rebuild(main/termux-apt-repo): with python 3.13

thunder-coding committed:
rebuild(x11/xapp): with python 3.13

thunder-coding committed:
rebuild(x11/mlt): with python 3.13

termux_setup_python_pip is called to ensure that cross-pip is used and
python library install path is detected properly by CMake, don't know
why this is needed when building with Python 3.13 on target, but not
when building with Python 3.12 on target. Maybe due to version mismatch
between host and target?

thunder-coding committed:
fix(main/python): hardcode android api level than detecting at runtime

[[TODO: add more descriptive commit message]]

thunder-coding committed:
rebuild(main/tizonia): with python 3.13

thunder-coding committed:
bump(main/python): 3.13.12

thunder-coding committed:
chore(main/python): allow on-device builds

ac_cv_func_getlogin_r=no is needed to ensure that it doesn't try to use
getlogin_r on Android during on-device builds.

Although I have enabled on-device builds, I don't see any point in
keeping this as the same version of python needs to be available for
building the same version of python. Pretty much defeats the purpose
unless you are rebuilding with some changes for the same minor version

thunder-coding committed:
rebuild(main/cinnamon): with python 3.13

thunder-coding committed:
chore(main/python): only disable getlogin_r for API level < 28

thunder-coding committed:
chore: Python 3.13 empty commit

All packages in this before this commit(s) before the last CI run will
be rebuilt manually using github workflow dispatch. This commit just
acts as a marker for reference so that I don't get lost while trying to
figure out which ones are left and also to not myself mad while trying
to figure out the dependency tree so that packages aren't built in the
wrong order.

%ci:no-build

vallentdavy1927 reviewed (commented):
Vallent
